File: rule-engine/src/main/java/com/frauddetection/domain/DataType.java
package com.frauddetection.domain;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.Id;
import org.springframework.data.relational.core.mapping.Column;
import org.springframework.data.relational.core.mapping.Table;
import java.time.Instant;
import java.util.UUID;

/**
 * DataType domain entity representing a schema for dynamic data types.
 * Defines the structure and allowed fields for a specific type of data.
 */
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
@Table("data_type")
public class DataType {
    @Id
    private UUID id;
    
    @Column("data_type")
    private String dataType;
    private String name;
    private String description;
    
    /**
     * JSON schema definition as a Map.
     * Example: {
     *   "fields": {
     *     "amount": "Double",
     *     "merchant": "String",
     *     "timestamp": "Instant"
     *   },
     *   "required": ["amount", "merchant"]
     * }
     */
    @Column("schema_definition")
    private String schemaDefinition; // JSON string
    
    @Column("sample_data")
    private String sampleData; // JSON string
    
    private Status status;
    
    @Column("created_at")
    private Instant createdAt;
    
    @Column("updated_at")
    private Instant updatedAt;
    
    @Column("created_by")
    private String createdBy;
    
    @Column("updated_by")
    private String updatedBy;
    
    public enum Status {
        ACTIVE, INACTIVE, DRAFT
    }
}

File: backend/internal/core/domain/data_type.go
package domain

import (
	"time"
)

// DataType represents a dynamic data type schema in the system
type DataType struct {
	ID               string    `json:"id" gorm:"type:uuid;primary_key;default:uuid_generate_v4()"`
	DataType         string    `json:"data_type" gorm:"column:data_type;uniqueIndex;not null"`
	Name             string    `json:"name" gorm:"not null"`
	Description      string    `json:"description" gorm:"type:text"`
	SchemaDefinition string    `json:"schema_definition" gorm:"column:schema_definition;type:jsonb"`
	SampleData       string    `json:"sample_data" gorm:"column:sample_data;type:jsonb"`
	Status           string    `json:"status" gorm:"not null;default:'ACTIVE'"`
	CreatedAt        time.Time `json:"created_at" gorm:"autoCreateTime"`
	UpdatedAt        time.Time `json:"updated_at" gorm:"autoUpdateTime"`
	CreatedBy        string    `json:"created_by" gorm:"column:created_by"`
	UpdatedBy        string    `json:"updated_by" gorm:"column:updated_by"`
}

// TableName specifies the table name for GORM
func (DataType) TableName() string {
	return "data_types"
}

// DataTypeStatus represents the possible statuses for a data type
const (
	DataTypeStatusActive   = "ACTIVE"
	DataTypeStatusInactive = "INACTIVE"
	DataTypeStatusDraft    = "DRAFT"
)

// CreateDataTypeRequest represents the request to create a data type
type CreateDataTypeRequest struct {
	DataType         string                 `json:"data_type" binding:"required"`
	Name             string                 `json:"name" binding:"required"`
	Description      string                 `json:"description"`
	SchemaDefinition map[string]interface{} `json:"schema_definition" binding:"required"`
	SampleData       map[string]interface{} `json:"sample_data"`
	Status           string                 `json:"status"`
	CreatedBy        string                 `json:"created_by"`
}

// UpdateDataTypeRequest represents the request to update a data type
type UpdateDataTypeRequest struct {
	DataType         *string                `json:"data_type,omitempty"`
	Name             *string                `json:"name,omitempty"`
	Description      *string                `json:"description,omitempty"`
	SchemaDefinition map[string]interface{} `json:"schema_definition,omitempty"`
	SampleData       map[string]interface{} `json:"sample_data,omitempty"`
	Status           *string                `json:"status,omitempty"`
	UpdatedBy        *string                `json:"updated_by,omitempty"`
}

// DataTypeListResponse represents the response for listing data types
type DataTypeListResponse struct {
	DataTypes []DataType `json:"data_types"`
	Total     int64      `json:"total"`
	Limit     int        `json:"limit"`
	Offset    int        `json:"offset"`
}

// DataTypeSearchRequest represents a search request for data types
type DataTypeSearchRequest struct {
	Search string `json:"search"`
	Status string `json:"status,omitempty"` // Optional filter by status
}
